### ISSUE TITLE
HDF5 pb (1.8.15-patch1) - H5Pcreate(): not a property list class

### ISSUE THREAD

**Issue by Gijom** [aldanor/hdf5-rs#10]
Hi,

I am trying to use the the library (cargo version) but fail with the HDF5 library version mentionned in the title (under archlinux).

The code I use is:
```
extern crate hdf5_rs;

fn main() {
    let mut matfile = hdf5_rs::File::open("/path/to/h5ex_t_int.h5", "r").unwrap();
    matfile.close();
}
```

The error I get is:
```
$ cargo run                                                                                                                  ✹ ✭[1853][18:12:47]
     Running `target/debug/loadmat`
HDF5-DIAG: Error detected in HDF5 (1.8.15-patch1) thread 140264758147072:
  #000: H5P.c line 299 in H5Pcreate(): not a property list class
    major: Invalid arguments to routine
    minor: Inappropriate type
thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: "H5Pcreate(): not a property list class"', src/libcore/result.rs:688
Process didn't exit successfully: `target/debug/loadmat` (exit code: 101)
```

BTW the cargo indicate v 0.1.0 while the git indicates v 1.5.0 is this normal or the potential souce of my problem ?

Thanks for any tips you could provide

**Comment by Gijom** [comment]
Just to confirm that on another machine with hdf5 v 1.8.12 hdf5-rs works perfectly. This might thus be related in recent changes in hdf5 lib.

**Comment by aldanor** [comment]
Hmm, have you tried running the test suite? (`cargo test`)

**Comment by Gijom** [comment]
So far I only used hdf5-rs as a dependency (with the error mentionned above on archlinux only).

So I cloned the repository and runned `cargo test`:
```
Compiling hdf5-rs v0.1.1 (file:///home/tabs/Sources/extern/hdf5-rs)
error: linking with `cc` failed: exit code: 1
note: "cc" "-Wl,--as-needed" "-m64" "-L" "/usr/lib64/rustlib/x86_64-unknown-linux-gnu/lib" "/home/tabs/Sources/extern/hdf5-rs/target/debug/hdf5_rs-ccd42abdc86e010a.0.o" "-o" "/home/tabs/Sources/extern/hdf5-rs/target/debug/hdf5_rs-ccd42abdc86e010a" "-Wl,--gc-sections" "-pie" "-nodefaultlibs" "-L" "/home/tabs/Sources/extern/hdf5-rs/target/debug" "-L" "/home/tabs/Sources/extern/hdf5-rs/target/debug/deps" "-L" "/usr/lib64/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "-Wl,-Bdynamic" "/home/tabs/Sources/extern/hdf5-rs/target/debug/deps/liblibhdf5_sys-3b201b50ea6a4836.rlib" "/home/tabs/Sources/extern/hdf5-rs/target/debug/deps/libregex-3e2d6a44b337590d.rlib" "/usr/lib/rustlib/x86_64-unknown-linux-gnu/lib/libtest-6a154fe0.rlib" "/usr/lib/rustlib/x86_64-unknown-linux-gnu/lib/libserialize-6a154fe0.rlib" "/usr/lib/rustlib/x86_64-unknown-linux-gnu/lib/libterm-6a154fe0.rlib" "/home/tabs/Sources/extern/hdf5-rs/target/debug/deps/libregex_syntax-3b2362e6ea0109f1.rlib" "/usr/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblog-6a154fe0.rlib" "/home/tabs/Sources/extern/hdf5-rs/target/debug/deps/libnum-d3b232b2ec3de579.rlib" "/home/tabs/Sources/extern/hdf5-rs/target/debug/deps/librustc_serialize-d9e72695d437325f.rlib" "/usr/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgetopts-6a154fe0.rlib" "/home/tabs/Sources/extern/hdf5-rs/target/debug/deps/liblazy_static-007034d2ad8108ce.rlib" "/home/tabs/Sources/extern/hdf5-rs/target/debug/deps/libremutex-7093d58b93d02587.rlib" "/home/tabs/Sources/extern/hdf5-rs/target/debug/deps/libaho_corasick-eea45523c8483aac.rlib" "/home/tabs/Sources/extern/hdf5-rs/target/debug/deps/libmemchr-48d9c161518e6940.rlib" "/home/tabs/Sources/extern/hdf5-rs/target/debug/deps/libtempdir-9fc76057b7d32c58.rlib" "/home/tabs/Sources/extern/hdf5-rs/target/debug/deps/librand-6fab17fc670a2bc8.rlib" "/home/tabs/Sources/extern/hdf5-rs/target/debug/deps/liblibc-d2268fc21ed63f2c.rlib" "/home/tabs/Sources/extern/hdf5-rs/target/debug/deps/libutf8_ranges-afda04de2a52d0c6.rlib" "/home/tabs/Sources/extern/hdf5-rs/target/debug/deps/libmempool-7f30f4503e1fecab.rlib" "/usr/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-6a154fe0.rlib" "/usr/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcollections-6a154fe0.rlib" "/usr/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_unicode-6a154fe0.rlib" "/usr/lib/rustlib/x86_64-unknown-linux-gnu/lib/librand-6a154fe0.rlib" "/usr/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-6a154fe0.rlib" "/usr/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc_jemalloc-6a154fe0.rlib" "/usr/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-6a154fe0.rlib" "/usr/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-6a154fe0.rlib" "-l" "hdf5" "-l" "dl" "-l" "pthread" "-l" "gcc_s" "-l" "pthread" "-l" "c" "-l" "m" "-l" "rt" "-l" "compiler-rt"
note: /home/tabs/Sources/extern/hdf5-rs/target/debug/hdf5_rs-ccd42abdc86e010a.0.o: In function `hdf5_rs::globals::os::H5P_LST_LINK_ACCESS_ID.::std::ops::Deref::deref::__static_ref_initialize':
/home/tabs/Sources/extern/hdf5-rs/src/globals.rs:62: undefined reference to `H5P_LST_LINK_ACCESS_g'
collect2: error: ld returned 1 exit status

error: aborting due to previous error
Could not compile `hdf5-rs`.
```

**Comment by aldanor** [comment]
Could you check the symbol names in your HDF5 library? E.g., on my OSX machine:

```bash
$ nm -a /usr/local/lib/libhdf5.10.dylib | grep LINK_ACCESS
000000000021b664 D _H5P_CLS_LINK_ACCESS_ID_g
000000000021d090 S _H5P_CLS_LINK_ACCESS_g
000000000021b6a0 D _H5P_LST_LINK_ACCESS_ID_g
```

The problem is -- on Windows/OSX these variables end with `_ID_g`, whereas on Linux they *normally* end with `_g` as far as I know. Hence this code: https://github.com/aldanor/hdf5-rs/blob/master/libhdf5-sys/src/h5p.rs#L114.

**Comment by aldanor** [comment]
Hmm, what's `target_os` set to on archlinux? (from the reference, all available options are: "windows", "macos", "ios", "linux", "android", "freebsd", "dragonfly", "bitrig" , "openbsd" or "netbsd")

**Comment by Gijom** [comment]
The command `nm -a /usr/lib64/libhdf5.so.10.0.1` returns `nm: /usr/lib64/libhdf5.so.10.0.1: no symbols`
Quite strange...
But `readelf -s /usr/lib64/libhdf5.so.10.0.1 | grep LINK_ACCESS` returns
```
1094: 00000000004c1eb8     8 OBJECT  GLOBAL DEFAULT   24 H5P_CLS_LINK_ACCESS_g
1552: 00000000004c0568     4 OBJECT  GLOBAL DEFAULT   23 H5P_LST_LINK_ACCESS_ID_g
 2094: 00000000004c05a4     4 OBJECT  GLOBAL DEFAULT   23 H5P_CLS_LINK_ACCESS_ID_g

```

I could not find how to get the `target_os` any idea ?

**Comment by aldanor** [comment]
So, the names end with `_ID_g` (like on OSX) and not with `_g` like on most Linux distros, that's the problem. The question is, how to figure it out in advance? (maybe the build script should do it?) Hmm...

**Comment by Gijom** [comment]
After a short google search I have the feeling it is more related to a version issue than a system issue:
https://www.hdfgroup.org/HDF5/doc/ADGuide/Compatibility_Report/CR_1.8.14.html

In short it seems tath the symbols have change on the 1.8.14  version of HDF5lib. H5P_LST_LINK_ACCESS_g has been removed and 5P_LST_LINK_ACCESS_ID_g added. Maybe the build test should be done on the library version.

**Comment by aldanor** [comment]
That's a good catch. I wonder how it should best be done -- because to build the crate, you need to know HDF5 version, but to get the version you need to link the code to libhdf5 first. Maybe there should be a mini build crate that does this intermediate step...

What official docs recommend is not very portable (e.g., wouldn't work on Windows) -- https://www.hdfgroup.org/hdf5-quest.html#vers1

**Comment by aldanor** [comment]
@Gijom could you please try running `cargo test` with both the old and the new version of HDF5 on archlinux using this branch: https://github.com/aldanor/hdf5-rs/tree/feature/conditional-build

Basically, there's another intermediate compilation step now which checks the actual version of the HDF5 library that's being linked and injects `#[cfg]` attributes into both `libhdf5_sys` and `hdf5_rs` crates -- this would be useful for other purposes as well, like enabling/disabling certain functionality based on HDF5 version.

**Comment by aldanor** [comment]
The related parts: 
- https://github.com/aldanor/hdf5-rs/commit/f7cedf0d87436f7f11e1e3b4020e139e8305a6e7
- https://github.com/aldanor/hdf5-rs/commit/150a4770cfd0cdbaf915c935018709372be9129e

**Comment by Gijom** [comment]
Hi, I tried `cargo test` with both the heads of master and feature/conditional-build on the "faulty" machine (the one with the recent version of libhdf5):
- master: fails with the same error than previously;
- features/conditional-build: compiles and all tests are successful.

I then compiled my short code with an overrided version of hdf5-rs and that works perfectly. Many thanks for you fast reaction. I think the last step it to merge master ?
